draft-ietf-ppm-dap.md

```diff
@@ -1200,34 +1182,120 @@ enum {
+splitting each report into report shares, one for each Aggregator. The Leader
+and Helper then run the aggregate initialization flow to accomplish two tasks:
+
+1. Recover and determine which input report shares are valid.
+1. For each valid report share, initialize the VDAF preparation process (see
+   {{Section 5.2 of !VDAF}}).
+
 The Leader and Helper initialization behavior is detailed below.
 
 #### Leader Initialization {#leader-init}
 
 The Leader begins the aggregate initialization phase with the set of candidate
-report shares as follows:
+reports as follows:
 
 1. Generate a fresh AggregationJobID. The ID value MUST be unique within the
    scope of the corresponding DAP task.
 1. Decrypt the input share for each report share as described in
    {{input-share-decryption}}.
 1. Check that the resulting input share is valid as described in
    {{input-share-validation}}.
-1. Initialize VDAF preparation as described in {{input-share-prep}}.
 
-If any step invalidates the report share, the Leader removes the report share
-from the set of candidate reports. Once the leader has initialized this state
-for all valid candidate report shares, it creates an `AggregationJobInitReq`
-message for each Helper to initialize the preparation of this candidate set. The
-`AggregationJobInitReq` message is structured as follows:
+If any step invalidates the report, the Leader rejects the report and removes
+it from the set of candidate reports.
+
+Next, for each report the Leader executes the following procedure:
+
+~~~
+state = VDAF.ping_pong_start(vdaf_verify_key,
+                             True,
+                             agg_param,
+                             report_id,
+                             public_share,
+                             plaintext_input_share.payload)
+if state != Rejected():
+  (state, outbound) = VDAF.ping_pong_req(agg_param, state, None)
+~~~
+
+where:
+
+* `vdaf_verify_key` is the VDAF verification key for the task
+* `agg_param` is the VDAF aggregation parameter provided by the Collector (see
+  {{collect-flow}})
+* `report_id` is the report ID
+* `public_share` is the report's public share
+* `plaintext_input_share` is the Leader's `PlaintextInputShare`
+
+The methods are defined in {{!VDAF}}. This process determines the initial
+per-report `state`, as well as the initial `outbound` message to send to the
+Helper. (These are coalesced into a single HTTP request to the Helper as
+described below.) If `state == Rejected()`, then the report is rejected and
+removed from the set of candidate reports.
+
+[[OPEN ISSUE: update references to the ping-pong API, once a VDAF draft is
+published which describes it.]]
+
+Next, for each candidate report the Leader constructs a `PrepareStep` message
+of type "initialize":
 
 ~~~
 struct {
-  ReportMetadata report_metadata;
-  opaque public_share<0..2^32-1>;
-  HpkeCiphertext encrypted_input_share;
-} ReportShare;
+  PrepareStepState prepare_step_state = 0; /* initialize */
```

Comment:
I like this notation, too, but since we refer to RFC 8446 section 3, I think we would need to explain that this document's usage of that presentation language differs in this one respect. I think the paragraph just before `{{overview}}` where we introduce the presentation language:

>This document uses the presentation language of {{!RFC8446}} to define messages
>in the DAP protocol. Encoding and decoding of these messages as byte strings
>also follows {{RFC8446}}.

would be a good place to add some text explaining our object literal notation.

Reply:
This notation is borrowed from [Section 7.2](https://datatracker.ietf.org/doc/html/rfc8446#section-7.1). I don't believe it's formalized anywhere.

Meta comment here: Let's not let perfection be the enemy of the good. We have a lot of work to do on the spec, and a this stage the most important thing is that it's unambiguous. If adding some text to clarify this notation would help, then go for it.

Reply:
I agree with Chris that this notation is good and we should use it. I scanned a handful of other documents in the TLS WG looking for prior art on this and couldn't find anything. I think our best path forward will be to add a paragraph or two to DAP explaining this notation. I wrote up https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/472 for us to clarify this notation.